doc/qbittorrent.1.md

```diff
@@ -1,36 +1,236 @@
+
+_ENVIRONMENT_ variables and command-line _OPTIONS_ can currently only override a small subset of all the available settings that can be configured in the main configuration file.
+
+qBittorrent can be cleanly terminated by fully closing the GUI (including the system tray, if that option is enabled) or by sending the **SIGINT** or **SIGTERM** signals.
 
 # OPTIONS
-**`--help`** Prints the command line options.
 
-**`--version`** Prints qbittorrent program version number.
+## Generic program information
+
+`-h`, `--help`
+
+:    Display a short help text about the command-line options and exit.
+
+`-v`, `--version`
+
+:    Display program version and exit.
+
+## General configuration
+
+`--configuration=`_NAME_
+
+:    Use a different configuration with name _NAME_. This enables running multiple separate instances of qBittorrent at the same time without conflict (provided they don't listen on the same ports for WebUI access, incoming connections, etc). Configuration and data files will be stored under a configuration hierarchy similar to the default one (see the _FILES_ section for more information) but with a final directory named **qBittorrent\__NAME_** instead of **qBittorrent**. For example, while the main configuration file is stored as _~/.config/qBittorrent/qBittorrent.conf_, the configuration file for a configuration named **foo** will be stored as _~/.config/qBittorrent\_**foo**/qBittorrent.conf_. If a configuration with the given _NAME_ does not exist yet, it is created with default settings. If it already exists, it is loaded as a normal configuration would be.
+
+`--no-splash`
+
+:    Do not display the splash screen on startup.
+
+`--profile=`_PATH_
+
+:    Store configuration _FILES_ in _PATH_. This enables what is known as "portable mode". Similarly to **`--configuration`**, this option can be used to run multiple separate instances of qBittorrent at the same time without conflitcts (the same caveats about listening port conflicts and other potential unavoidable conflicts apply). The difference is that configuration files are all stored together under the specified configuration _PATH_, instead of scattered under a separate directory hierarchy in the system's configuration directories (see the _FILES_ section for more information). This enables one to migrate their entire configuration from system to system by simply copying around one directory. If _PATH_ does not exist, it is created, with configuration files initialized with default settings inside it that a new instance of qBittorrent loads. If it exists, the configurations are loaded from there. A relative _PATH_ may be used, in which case it will be relative to the current working directory that qBittorrent is started from. Note that **`~`** will not get expanded, and will be interpreted as a folder literally named **`~`**. If both **`--configuration`** and **`--profile`** are specified, only **`--profile`** takes effect. Implies **`--relative-fastresume`**.
```

Comment:
>This enables what is known as "portable mode".

It is incorrect. It just changes profile directory path.

Reply:
@glassez 
Then how exactly is "portable mode" enabled?

Maybe I should change that sentence to read `This functionally enables what is known as "portable mode".`?

When I invoked qbittorrent with this option, it created a standalone config/data file directory. When I started it again with the same path as argument, it used those settings it created before. Isn't that what "portable mode" is? You could send the folder to a friend and they would have the exact same config as you.

Reply:
>You could send the folder to a friend and they would have the exact same config as you.

How about existing torrents?
>Then how exactly is "portable mode" enabled?

What does "portable mode" mean exactly?
IMO, it should allow you to take some folder containing the application binary and all its configuration files, transfer it to another computer, and run it without any additional effort. It doesn't seem to apply to qBittorrent.
I never understood how qBittorrent could be really portable. Even currently I would remove such "feature".

Reply:
@glassez 
> How about existing torrents?

Good point. I'll change the text to mention this and it's implications.

>What does "portable mode" mean exactly?
IMO, it should allow you to take some folder containing the application binary and all its configuration files, transfer it to another computer, and run it without any additional effort. It doesn't seem to apply to qBittorrent.

It's mostly a Windows think. As far as I know, in Windows land, a "portable application"/"portable mode" is almost a synonym of "static build", but it provides some additional guarantees. Not only can it run from a standalone directory without depending on external libraries, it also stores its settings within the standalone directory. It does not store config files in %APPDATA% or use/set registry keys. Thus, the application directory can be shared freely between systems, and it will work with exactly the same options.

From what i can observe, using `--profile` makes qBIttorrent fit this description, at least on Windows. The official build is already static, and the `--profile` flag can be used to keep all the configuration/data files in any directory, including the same one as the binary.

We could even provide an official "portable" release for Windows - just a zipped folder with the files you'd usually find in an installation folder (`.exe`, `.pdb`, ...), plus a `qbittorrent.bat` folder that just does `qbittorrent.exe --profile=.\`.

But it can also be useful for other platforms. For example, imagine you are trying to debug some issue and the user is using portable mode; they send you their "profile" folder, you run qBittorrent with `--profile` and the path to the user's profile and boom, you have a similar setup to that of the user just like that. Or you could even just ask the user for the .conf, construct a profile folder manually, with that, and execute qbittorrent with `--profile`.

So I think this it is indeed accurate to say "This functionally enables what is known as "portable mode". But let me know if I misunderstood your point or some other detail.

Reply:
>As far as I know, in Windows land, a "portable application"/"portable mode" is almost a synonym of "static build"

It is incorrect statement. Most of dynamically linked Windows applications come with their own library binaries (except the Windows libraries and maybe VC runtime) so it is not so hard to turn it to so-called "portable mode".
But it's off-topic here.
>but it provides some additional guarantees. Not only can it run from a standalone directory without depending on external libraries, it also stores its settings within the standalone directory. It does not store config files in %APPDATA% or use/set registry keys. Thus, the application directory can be shared freely between systems, and it will work with exactly the same options.

It's almost the same as I said before.
The main thing is when user has "portable" app it should be able to `run it without any additional effort` (as I said above).
>From what i can observe, using --profile makes qBIttorrent fit this description, at least on Windows.

No. This option is the one that can help to provide "portable mode" but it isn't enough.
The another one is "--relative-fastresume" that allows to transfer existing torrents with the app. But unfortunately it isn't enough too since it can't cover all the cases so user is still responsible for some additional cares.
>But it can also be useful for other platforms. [...]

It is true. The "--profile" option itself is good for some purposes. But it only does what it does (it just "allows you to specify custom profile path"). I don't want you to identify it with anything more.
>We could even provide an official "portable" release for Windows - just a zipped folder with the files you'd usually find in an installation folder (.exe, .pdb, ...), plus a qbittorrent.bat folder that just does qbittorrent.exe --profile=.\.

Really you don't need to have any .bat. If you create "profile" folder near the application executable qBittorrent will detect it and enables "relative fastresume" for you.
>But let me know if I misunderstood your point or some other detail.

Just call a spade a spade and don't give users too much hope. Otherwise, we will have to be responsible for it.

I don't want to maintain "portable mode".
If someone want to do it you can ignore my comments and follow "portable mode maintainer" thoughts.

@Chocobo1, what do you say?

Reply:
@glassez 

>No. This option is the one that can help to provide "portable mode" but it isn't enough.
The another one is "--relative-fastresume" that allows to transfer existing torrents with the app. But unfortunately it isn't enough too since it can't cover all the cases so user is still responsible for some additional cares.

`--profile` already implies `--relative-fastresume`. What are the "additional cases" that you refer to?

> If you create "profile" folder near the application executable qBittorrent will detect it and enables "relative fastresume" for you.

You mean a folder literally named `profile`? I did not know of this.

> I don't want to maintain "portable mode".
If someone want to do it you can ignore my comments and follow "portable mode maintainer" thoughts.

I understand not wanting to maintain it, and I'm not pushing for it. I'm just trying to document the existing behavior/features as best as I can. In light of your comments, it is clear I need to reword the text to not mention "portable" mode explicitly, I just need to know more details about what I'm asking here.

Reply:
>--profile already implies --relative-fastresume.

What makes you think that?

Reply:
@glassez 

https://github.com/qbittorrent/qBittorrent/blob/a10cee0f2be2b57b8550e5a02cbeb6ae703cd910/src/app/application.cpp#L160

and

https://github.com/qbittorrent/qBittorrent/blob/a10cee0f2be2b57b8550e5a02cbeb6ae703cd910/src/app/application.cpp#L184

Reply:
You misunderstood it...

Reply:
@glassez How so? Can you please clarify then? This is starting to get really confusing.

Reply:
Your example relates to the following:
>If you create "profile" folder near the application executable qBittorrent will detect it and enables "relative fastresume" for you.

Reply:
@glassez 
So let's get this straight:

- Specifying `--profile=some/path` with a name of some folder does not imply `--relative-fastresume`, so the BT_backup folder will not be portable between computers
- Specifying `--profile=some/path` and `--relative-fastresume` will make the BT_backup folder portable between computers
- Creating a folder named `profile` next to the application executable is effectively the same as using both `--profile=profile` and `--relative-fastresume`
- In all three cases all other data/config files/folders are portable between computers.

Is this correct?

Reply:
Additionally: an instance started as above will be able to work alongside another instance with a different `--profile` or `--configuration` so long as they don't listen on the same ports and things like that. Yes?

Reply:
>Specifying `--profile=some/path` with a name of some folder does not imply `--relative-fastresume`, so the `BT_backup` folder will not be portable between computers

Yes.
>Specifying `--profile=some/path` and `--relative-fastresume` will make the BT_backup folder portable between computers

Yes. But it will not work in case of path that cannot be represented in relative form (e.g. UNC or path in different disk on Windows).
>Creating a folder named `profile` next to the application executable is effectively the same as using both `--profile=profile` and `--relative-fastresume`

Yes.
And it won't work if you set another profile path via command line.
>In all three cases all other data/config files/folders are portable between computers.

AFAIK, other data isn't affected at all. So there can be conflicts if you run it on different computer.
>an instance started as above will be able to work alongside another instance with a different --profile or --configuration so long as they don't listen on the same ports and things like that. Yes?

Seems Yes. At least it should...

Reply:
@glassez I have changed the documentation according to our discussion. Is it accurate now?
Also some more questions:

> Yes. But it will not work in case of path that cannot be represented in relative form (e.g. UNC or path in different disk on Windows).

What is "path" referring to here? the one passed to the `--profile` option or the paths in the `.fastresume`s themselves?

> Yes.
And it won't work if you set another profile path via command line.

What exactly won't work? Is qBittorrent not going to launch at all, or will it launch but not use the `profile` folder next to it as the profile?

---

Can `--relative-fastresume` be passed in without `--profile`? if so, what will the paths be relative to? The current working dir?

If some torrents are initially saved with absolute paths, but then I change all config options to use relative paths and run qBittorrent with `--relative-fastresume`, will all the existing fastresumes be updated to use relative paths?

Reply:
> > Yes. But it will not work in case of path that cannot be represented in relative form (e.g. UNC or path in different disk on Windows).
> 
> What is "path" referring to here? the one passed to the `--profile` option or the paths in the `.fastresume`s themselves?

In `.fastresume`. I can't imagine how it will react to such a path in the `--profile` option...
> > Yes.
> > And it won't work if you set another profile path via command line.
> 
> What exactly won't work? Is qBittorrent not going to launch at all, or will it launch but not use the `profile` folder next to it as the profile?

If profile path is explicitly set via command line it wiil not perform "profile" folder autodetection.
> Can `--relative-fastresume` be passed in without `--profile`? if so, what will the paths be relative to? The current working dir?

I don't know. Try to find out for yourself from the code, if you want...
> If some torrents are initially saved with absolute paths, but then I change all config options to use relative paths and run qBittorrent with `--relative-fastresume`, will all the existing fastresumes be updated to use relative paths?

I'm sorry, I don't know the answers to all these questions.
I am annoyed by this topic. I always thought all these "portable mode", "relative fastresume" and so on were unacceptable and I was against it, so I don't want to be responsible for it. I am sure that it is implemented extremely superficially and unreliably, and in many cases it has undefined behavior.
Please don't rape me any more with this topic.

Reply:
@glassez I'm sorry about the question spam, I just want to make sure I get everything right, to have a quality man page. I think I have enough information now, the rest I can figure out on my own if needed.

I agree that this kind of stuff is implemented in a weird and finicky way - there seem to be a lot of edge cases and it is hard to explain them all. On top of this, the gain for having these features is dubious and it is not clear what kind of use case really needs this. qBittorrent's installer does not do any magic behind the scenes or pollute the registry or whatever - it just uses a bunch of config files in standard and predictable locations. `--configuration` is enough by itself IMO. But until we all decide to remove `--profile` and friends, I think they should be documented as accurately as possible.

Reply:
>But until we all decide to remove `--profile` and friends,

Only "friends".
 `--profile` isn't bad for itself. It can be used in testing purposes.
>`--configuration` is enough by itself IMO.

I think No. At least on Windows different kind of configuration/data files are placed inside different folders (I mean there are at least two qBittorrent folders in different places). So `--profile` allows you to keep them together. BTW, it seems `--configuration` works incorrectly with default profile (on Windows). For me `--configuration=test` creates "cache_test" inside "qBittorrent" instead of "qBittorrent_test/cache" as I expect.

Reply:
@glassez 
> > But until we all decide to remove `--profile` and friends,
> 
> Only "friends".
> `--profile` isn't bad for itself. It can be used in testing purposes.

Fair enough.

> > `--configuration` is enough by itself IMO.
> 
> I think No. At least on Windows different kind of configuration/data files are placed inside different folders (I mean there are at least two qBittorrent folders in different places). So `--profile` allows you to keep them together.

If you mean "different folders" in the same way that on Linux the config file and the logs are in different folders, I don't think that's a problem in and of itself. Despite being different directories, they are _expected_; one would not expect the logs to be in `~/config`, after all. But like you said above, it's useful to keep everything together for testing purposes.

> BTW, it seems `--configuration` works incorrectly with default profile (on Windows). For me `--configuration=test` creates "cache_test" inside "qBittorrent" instead of "qBittorrent_test/cache" as I expect.

Does this _only_ happen with the `cache` folder? Either way, I think it's ok to document the intended behavior, any deviation from it is another issue/PR.

Reply:
@glassez 
I figured out the answers to the 2 outstanding questions above:

> Can `--relative-fastresume` be passed in without `--profile`?
> if so, what will the paths be relative to? The current working dir?

Yes it can, and the path will be relative to the path specified by `$HOME` (usually the home directory).

>If some torrents are initially saved with absolute paths, but then I change all config options to use relative paths and run qBittorrent with `--relative-fastresume`, will all the existing `.fastresume`s be updated to use relative paths?

No. Torrents previously added with absolute paths keep the absolute paths in the `.fastresume`s.

Reply:
@glassez I think I addressed everything about the `--configuration`, `--profile`, and `--relative-fastresume` options now. Please confirm. If everything looks good, I can continue the copy-paste to the nox manpage and make the final adjustments to it. Then I'll finally submit the PR for final review.

Reply:
@glassez I know you don't like reviewing text walls, but it would be great if you could please address my comment above. It's just a small portion of the overall PR.

Reply:
>Please confirm. 

Now it looks more like the truth (as far as I can tell about it).
The only thing I don't like is that you put "running multiple instances" as the main purpose of both `--configuration` and `--profile` options (at least it looks like this due to first place in option description).

Reply:
@glassez Thanks for reviewing, looks like we're getting closer.

> The only thing I don't like is that you put "running multiple instances" as the main purpose of both `--configuration` and `--profile` options (at least it looks like this due to first place in option description).

Well, what would you say is the main purpose of each option? Just supporting different configs for different scenarios?

If so, I could change it to something along the lines of: "These options enable saving and using different conigurations. Additionally, they can be leveraged to run multiple different instances of qBittorrent at the same time. However, in this case, one must take care to adjust settings that are inherently incompatible across multiple simultaneous instances if set to the same value, such as the listening port.". Is this what you're looking for?

Reply:
>Is this what you're looking for?

Apparently, Yes. But I can't confirm that you describe it in a better way. My English is too bad for that, and you are too verbose and informal for me to understand exactly everything...

>Well, what would you say is the main purpose of each option? Just supporting different configs for different scenarios?

You can just prepend your current description with something like this one (I mean to say about "supporting different configs for different scenarios").

Reply:
@glassez 

> Apparently, Yes. But I can't confirm that you describe it in a better way. My English is too bad for that, and you are too verbose and informal for me to understand exactly everything...

I've pushed the change, I think it conveys the intended idea/usage case better now.

About formality, was that comment directed at my example/sketch in the comment, or at the state of the PR in general? I'm trying to be as objective as possible with a reasonable level of formality for a man page. If you think the current text/tone is too informal/not objective enough (or too formal), can you please point to a specific instance (or several) and explain how you would fix it, and your reasoning for such changes?

Reply:
Sorry, I have nothing to add. I just wanted to say that it is easier for me to understand strictly formalized wordings because of my poor English.